apis/kueue/v1alpha1/tas_types.go

```diff
@@ -27,3 +53,57 @@ const (
 	// The label is set when starting the Job, and removed on stopping the Job.
 	PodSetLabel = "kueue.x-k8s.io/podset"
 )
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+	// TopologyLevels define the levels of topology.
+	//
+	// +listType=atomic
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=5
+	Levels []TopologyLevel `json:"levels,omitempty"`
+}
+
+// TopologyLevel defines the desired state of TopologyLevel
+type TopologyLevel struct {
+	// NodeLabel indicates the name of the node label for a specific topology
+	// level.
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=316
+	// +kubebuilder:validation:Pattern=`^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$`
+	NodeLabel string `json:"nodeLabel,omitempty"`
```

Comment:
```suggestion
	NodeLabel string `json:"nodeLabel"`
```

I guess that the `omitempty` marker makes this field an optional field, right?

Reply:
Ah, got it. Since the field is required the omitempty does not matter. I removed it for clarity.

Reply:
My concerns was related to https://github.com/kubernetes-sigs/controller-tools/issues/943.
So, before controller-tools v0.16.0, the `+required` marked is not respected when the field has `omitempty`.
But, the latest controller-tools allow us to make the omitempty field as required.

However, we decided to remove `omitempty` since we should not add `omitempty` to the required field as the API definition principal.